dockerized rails draft post with tech stack info

vijayabharathib opened the pull request:
Draft with first three sections: `What We Will Build`, `How We Will Build`, and `Environment`

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
```diff
@@ -0,0 +1,113 @@
+
+Here is the problem. See if you recognize it. You started work on a new project that needed a certain environmental setup or wanted to try out a new technology. You followed the guide to the letter and ran into issues launching the local application. No one else seems to have got the issue. **It works on their machine, but not on yours**. You consulted *stackoverflow* trying to figure out what we ...
+
+You don't recognize that? Lucky you. For the rest of us, `docker` solves the problem by allowing us to bundle our runtime environment in an image and share it around, just like a `git` repository.
```
Ok, here is the thing, when you are referring to a Proper noun, avoid putting it into a inline code block: `docker`. That is, use Docker and Git over here.

When you are referring to a command or some code, then you can use `git` or `docker`. See the difference?

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
```diff
@@ -0,0 +1,113 @@
+
+You don't recognize that? Lucky you. For the rest of us, `docker` solves the problem by allowing us to bundle our runtime environment in an image and share it around, just like a `git` repository.
+
+Sometimes, it just boils down to having a simple text file named `Dockerfile` with a set of commands to get the right environment.
```
Yes, `Dockerfile` works here as well. It is the name of a file that the user will have to create, so highlighting it with the inline code block is good.

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
```diff
@@ -0,0 +1,113 @@
+
+When I say environment, it is everything required to run a particular software. Docker images are solved problems for you to build your solution on top of.
+
+Let's say you need `ruby` runtime on `ubuntu`. Pull an `ubuntu` image and install `ruby` on top. Wait, isn't that what we do already? Yes. So, here is one better. How about pulling an image that already has `ruby` on top of `ubuntu`. Now we are talking? How about pulling an image that has `ruby` and `nodejs` on top of a `debian` flavor? Sweet.
```
Ruby instead of `ruby`, Ubuntu instead of `ubuntu`, ...

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
```diff
@@ -0,0 +1,113 @@
+
+Now, think about other solutions that solve similar problem. I know I've used a `dual-boot` PC that runs both Windows and Linux. I switch between them when required, but wouldn't be able to run both at the same time.
+
+Another solution is the virtual box that allows you to run entirely full scale guest OS on top of your host OS.
```
VirtualBox instead of virtual box or Virtual Box

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
````diff
@@ -0,0 +1,113 @@
+```Dockerfile
+FROM ruby:2-stretch
+
+```
````
blank line after code blocks

vijayabharathib commented:
Hi @brunokrebs , Fixes for the initial comments on the first half of the posts are included. In addition, rest of the workflow has also been written. Few deviations from the outline are details in **slack** channel, along with request for image uploads. Please have a look at them when you get a chance. thanks!

vijayabharathib commented:
Added another commit, fixed and re-ordered a few things while trying to follow my own steps. Created a whole new repository from scratch. All goes very smooth now.

vijayabharathib commented:
Part 2 has progressed significantly with the auto deployment workflow between GitHub, Travis & Heroku complete. Last mile of the run to finish the app.

vijayabharathib commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
Yes, I'm serious. :-) The ruby way (bundler) is quite opposite to npm. `Gemfile` is edited directly to make necessary changes and then `bundle install` brings in the binaries and dependencies & versions are locked to `Gemfile.lock` (the piece of functionality picked up by yarn now) so that every developer will get the [same version of the environment](http://bundler.io/rationale.html)  (almost close to what docker does for rest of the environment). 

Unlike npm, all gems are global unless you direct it to local (hence most gems are reused if they have been installed for a previous project). This is not because we are inside a container, that's generally the approach.

In fact, there is even a gem to bring in that command line capability if one wanted : [https://github.com/DruRly/gemrat](https://github.com/DruRly/gemrat). Doesn't look like the community adopted it much wider.

vijayabharathib commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
Yes, I sent these on March 18th on slack channel. I also have sent a few more earlier on Feb 14th. would be very useful to get URLs for them from Auth0 blog. (I'll have to replace the URL shown here, they are just placeholder for the images)

brunokrebs replied:
I have uploaded the following images:

- https://cdn.auth0.com/blog/docker-ruby/guard-browser-plugin-connection.png
- https://cdn.auth0.com/blog/docker-ruby/docker-guard-connections.gif
- https://cdn.auth0.com/blog/docker-ruby/automatic-test-runs.gif
- https://cdn.auth0.com/blog/docker-ruby/livereload.gif
- https://cdn.auth0.com/blog/docker-ruby/you-are-own-rails.png
- https://cdn.auth0.com/blog/docker-ruby/show_travis_ci_running_tests.gif
- https://cdn.auth0.com/blog/docker-ruby/starting_docker_container_local_server.gif
- https://cdn.auth0.com/blog/docker-ruby/show_heroku_deployment_after_travis_test.gif

After uploading, I realised that `show_travis_ci_running_tests.gif` has almost 4mb. I think we should avoid using it as our blog simply loads all assets when readers access it (no lazy loading).

vijayabharathib commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
There is an option to introduce [rack middleware gem](https://github.com/johnbintz/rack-livereload). Looks a bit outdated and not actively maintained. I tried it anyways, ran into an issue with port binding which is an open issue already. Will give it some more time to see if I could fix that issue.

vijayabharathib commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
You are right, on its own, the container is harmless. But it is possible (in fact, easy) to expose host root directory to the container as a volume and one can delete the whole set of files in host root wiping out everything. I've added some text on these lines now. You can have a look once I commit them to a pull request.

vijayabharathib commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
You are spot on. All commands create a layer. Most of the temporary layers are discarded in the final image except those created by `RUN`, `COPY` and `ADD`. I've added that now.

vijayabharathib commented:
I've left responses to some of the queries. For rest of them, I've updated the post to reflect your suggestions. Also ran it through Grammarly.

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
tell users that they can use `dockerized-rails-app`, naming stuff is boring and this will help them 😄

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
`/app` or `/project`? mine shows `/project`

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
`environment` <= it's plural actually

vijayabharathib replied:
these two files were not used. You can delete them if that would make sense to save some memory: `show_travis_ci_running_tests.gif` and `docker-guard-connections.gif`

`docker-guard-connections.gif` was redundant and out of place as it is covered in other gifs too.

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
where do I add it? right before `end`? please, provide details

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
I have never used `docker-compose` and every time I trigger it it gets "stuck" on the logs. So, I always wonder if I should hit ctrl + c to run other commands (like `docker-compose exec` on line 455) or if I open a new terminal. Can you add some instructions on the beginning of the article about it?

brunokrebs replied:
For some reason I got the following error and ended up having to `rm ./tmp/pids/server.pid` from project root.

brunokrebs replied:
I just figured that when I run `docker-compose down` something goes silently wrong and I have to `rm ./tmp/pids/server.pid` before `docker-compose up` again.

If you don't know the reason, you can just add a note somewhere about this.

brunokrebs replied:
hehehe, ok, I can see you added it on line 488, perhaps you should move it somewhere else

vijayabharathib replied:
Hi, I actually fixed the whole issue by introducing a temporary volume to the `docker-compose.yml` file. and yes, moved the section to be part of the `foreman` introduction. This `server.pid` shouldn't be an issue anymore. You'll be able to see this when I push the changes.

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
it worked 😄

brunokrebs commented on articles/2018-02-27-dockerized-workflow-ruby-on-rails-auth0.markdown:
hehehe, ok, I can see you added it here

vijayabharathib commented:
Hi Bruno, part 2 is also updated now. I've upgraded that to 5.2 version. In fact, part 1 introductory paragraph also been updated to reflect new repository and applications. Both ready for your review.
